File: Sources/Sideros/Models/Qwen3.swift
import Foundation
import MLX

public struct Qwen3Activations {
    public let embeddings: MLXArray
    public let blocks: [MLXArray]
    public let norm: MLXArray
    public let logits: MLXArray
}

/// Qwen2 with three deltas: q/k are rms-normed per head between the projection and the
/// rotation, the projections carry no bias, and heads * headDim is decoupled from
/// hiddenSize (the small sizes widen, o_proj narrows back).
public struct Qwen3: CausalLM {
    private let config: Qwen3Config
    private let parameters: [String: MLXArray]
    private let quantization: Quantization?
    private let skinny: Bool
    /// One compiled closure per layer: the trace fuses silu(gate) · up into one kernel.
    let mlps: [@Sendable (MLXArray) -> MLXArray]

    public init(
        config: Qwen3Config, parameters: [String: MLXArray], quantization: Quantization? = nil,
        skinny: Bool = false
    ) {
        self.config = config
        self.quantization = quantization
        self.skinny = skinny

        // One fused projection per layer instead of three, as in Qwen2 — row-aligned in
        // every representation. No bias to fuse here.
        var parameters = parameters
        for layer in 0..<config.numHiddenLayers {
            let p = "model.layers.\(layer).self_attn."
            for suffix in ["weight", "scales", "biases"] {
                let parts = ["q_proj.", "k_proj.", "v_proj."].compactMap {
                    parameters[p + $0 + suffix]
                }
                guard parts.count == 3 else { continue }
                parameters[p + "qkv_proj." + suffix] = concatenated(parts, axis: 0)
            }
        }
        self.parameters = parameters

        mlps = (0..<config.numHiddenLayers).map { layer in
            let p = "model.layers.\(layer).mlp."
            let gate = Linear(
                p + "gate_proj", parameters: parameters, quantization: quantization, skinny: skinny)
            let up = Linear(
                p + "up_proj", parameters: parameters, quantization: quantization, skinny: skinny)
            let down = Linear(
                p + "down_proj", parameters: parameters, quantization: quantization, skinny: skinny)
            // silu spelled out: MLXNN's is itself a compiled function with its own lock,
            // and entering it while this trace holds the global eval lock inverts the
            // order against any thread calling it directly — an ABBA deadlock.
            return compile { x in
                let gated = gate(x)
                return down(gated * sigmoid(gated) * up(x))
            }
        }
    }

    /// Total by construction: loadQwen3Parameters rejects any checkpoint missing a name.
    private subscript(name: String) -> MLXArray { parameters[name]! }

    private func linear(_ name: String) -> Linear {
        Linear(name, parameters: parameters, quantization: quantization, skinny: skinny)
    }

    /// No positional embedding table: position enters only through RoPE.
    public func embed(_ tokens: MLXArray, offset: Int = 0) -> MLXArray {
        let table = Embedding("model.embed_tokens", parameters: parameters, quantization: quantization)
        return table(tokens)
    }

    public func makeCache() -> [KVCache] { (0..<config.numHiddenLayers).map { _ in KVCache() } }

    public func logits(_ embeddings: MLXArray, cache: [KVCache])
        -> (logits: MLXArray, cache: [KVCache])
    {
        let decoded = decode(embeddings, cache: cache)
        return (decoded.logits, decoded.cache)
    }

    public func callAsFunction(_ ids: [Int32]) -> Qwen3Activations {
        let embeddings = embed(ids)
        let decoded = decode(embeddings, cache: makeCache())
        return Qwen3Activations(
            embeddings: embeddings,
            blocks: decoded.blocks,
            norm: decoded.norm,
            logits: decoded.logits
        )
    }

    private func decode(_ embeddings: MLXArray, cache: [KVCache])
        -> (blocks: [MLXArray], norm: MLXArray, logits: MLXArray, cache: [KVCache])
    {
        var cache = cache
        var x = embeddings
        var blocks: [MLXArray] = []
        for layer in 0..<config.numHiddenLayers {
            (x, cache[layer]) = block(x, layer: layer, cache: cache[layer])
            blocks.append(x)
        }

        let normed = rmsNorm(x, prefix: "model.norm")
        let head = config.tieWordEmbeddings ? "model.embed_tokens" : "lm_head"
        return (blocks, normed, linear(head)(normed), cache)
    }

    func block(_ x: MLXArray, layer: Int, cache: KVCache) -> (MLXArray, KVCache) {
        let prefix = "model.layers.\(layer)."
        let (context, grown) = attention(
            rmsNorm(x, prefix: prefix + "input_layernorm"), prefix: prefix, cache: cache)
        let attended = x + context
        return (
            attended
                + mlps[layer](rmsNorm(attended, prefix: prefix + "post_attention_layernorm")),
            grown
        )
    }

    func attention(_ x: MLXArray, prefix: String, cache: KVCache) -> (MLXArray, KVCache) {
        let length = x.dim(1)
        let offset = cache.offset
        let p = prefix + "self_attn."

        let queryWidth = config.numAttentionHeads * config.headDim
        let keyValueWidth = config.numKeyValueHeads * config.headDim
        let fused = linear(p + "qkv_proj")(x)
        let parts = split(fused, indices: [queryWidth, queryWidth + keyValueWidth], axis: -1)

        func heads(_ part: MLXArray, _ count: Int) -> MLXArray {
            part.reshaped(1, length, count, config.headDim).transposed(0, 2, 1, 3)
        }

        // The norm weight is one headDim vector shared by every head and position.
        let queries = rope(
            rmsNorm(heads(parts[0], config.numAttentionHeads), prefix: p + "q_norm"),
            offset: offset)
        let (keys, values, grown) = cache.appending(
            keys: rope(
                rmsNorm(heads(parts[1], config.numKeyValueHeads), prefix: p + "k_norm"),
                offset: offset),
            values: heads(parts[2], config.numKeyValueHeads))

        // GQA: the kernel broadcasts the kv heads across the query heads itself. See
        // Qwen2 for why `.causal` and why a single query goes unmasked.
        let attended = MLXFast.scaledDotProductAttention(
            queries: queries, keys: keys, values: values,
            scale: 1 / sqrt(Float(config.headDim)), mask: length == 1 ? .none : .causal)

        let projected = linear(p + "o_proj")(
            attended.transposed(0, 2, 1, 3).reshaped(1, length, queryWidth))
        return (projected, grown)
    }

    func rmsNorm(_ x: MLXArray, prefix: String) -> MLXArray {
        MLXFast.rmsNorm(x, weight: self[prefix + ".weight"], eps: config.rmsNormEps)
    }

    /// `traditional: false` is the interleaving HF calls `rotate_half`: the vector is
    /// split in two halves rather than into adjacent pairs.
    func rope(_ x: MLXArray, offset: Int = 0) -> MLXArray {
        MLXFast.RoPE(
            x, dimensions: config.headDim, traditional: false, base: config.ropeTheta,
            scale: 1, offset: offset)
    }
}

File: packages/sideros/src/sideros/models/qwen3.py
"""Qwen3 dense: Qwen2 with per-head q/k RMSNorm, no bias, explicit head_dim.

Authoritative semantics: transformers' modeling_qwen3.py. The norms sit between the
projections and the rotation, and `num_attention_heads * head_dim` is decoupled from
`hidden_size` (0.6B: 16x128 = 2048 against a 1024 trunk).

Two load-time fusions, both on the output axis (row-aligned in every representation,
so dense and packed u32 fuse the same way): q‖k‖v into `qkv_proj` and gate‖up into
`mlp.gate_up_proj`. The tree declares only the fused names.
"""

import math
from dataclasses import dataclass
from typing import NamedTuple

import mlx.core as mx
import mlx.nn as nn

from sideros.core.cache import KVCache
from sideros.core.kernels.rope_epilogue import rope_epilogue, rope_epilogue_applies
from sideros.core.layers import SwiGLU, split_qkv


@dataclass(frozen=True)
class Qwen3Config:
    hidden_size: int
    num_hidden_layers: int
    num_attention_heads: int
    num_key_value_heads: int
    head_dim: int
    vocab_size: int
    rms_norm_eps: float
    rope_theta: float
    tie_word_embeddings: bool
    intermediate_size: int
    quantization: tuple[int, int] | None  # (group_size, bits)


class Qwen3Attention(nn.Module):
    def __init__(self, config: Qwen3Config) -> None:
        super().__init__()
        self.heads = config.num_attention_heads
        self.kv_heads = config.num_key_value_heads
        self.head_dim = config.head_dim
        self.rope_theta = config.rope_theta
        self.eps = config.rms_norm_eps
        hidden = config.hidden_size
        queries = self.heads * self.head_dim
        key_values = self.kv_heads * self.head_dim
        self.qkv_proj = nn.Linear(hidden, queries + 2 * key_values, bias=False)
        self.o_proj = nn.Linear(queries, hidden, bias=False)
        self.q_norm = nn.RMSNorm(self.head_dim, eps=config.rms_norm_eps)
        self.k_norm = nn.RMSNorm(self.head_dim, eps=config.rms_norm_eps)

    def split_heads(self, x: mx.array) -> tuple[mx.array, mx.array, mx.array]:
        """qkv split back into per-head [1, heads, length, head_dim], normed but
        unrotated — the boundary transformers' q_norm/k_norm hooks expose."""
        q, k, v = split_qkv(
            self.qkv_proj(x),
            heads=self.heads,
            kv_heads=self.kv_heads,
            head_dim=self.head_dim,
        )
        return self.q_norm(q), self.k_norm(k), v

    def rope(self, x: mx.array, offset: int) -> mx.array:
        return mx.fast.rope(
            x, self.head_dim, traditional=False, base=self.rope_theta, scale=1.0, offset=offset
        )

    def step_heads(self, x: mx.array, offset: int) -> tuple[mx.array, mx.array, mx.array]:
        """T=1: q/k norm + rotation in one dispatch, v a free slice of the projection."""
        fused = self.qkv_proj(x)
        queries, keys = rope_epilogue(
            fused,
            query_heads=self.heads,
            kv_heads=self.kv_heads,
            head_dim=self.head_dim,
            q_norm=self.q_norm.weight,
            k_norm=self.k_norm.weight,
            offset=offset,
            base=self.rope_theta,
            eps=self.eps,
        )
        values = fused[..., (self.heads + self.kv_heads) * self.head_dim :]
        return (
            queries.reshape(1, self.heads, 1, self.head_dim),
            keys.reshape(1, self.kv_heads, 1, self.head_dim),
            values.reshape(1, self.kv_heads, 1, self.head_dim),
        )

    def __call__(self, x: mx.array, cache: KVCache) -> mx.array:
        length = x.shape[1]
        offset = cache.offset
        if length == 1 and rope_epilogue_applies(self.head_dim):
            queries, rotated, v = self.step_heads(x, offset)
        else:
            q, k, v = self.split_heads(x)
            queries, rotated = self.rope(q, offset), self.rope(k, offset)
        keys, values = cache.update_and_fetch(rotated, v)
        attended = mx.fast.scaled_dot_product_attention(
            queries, keys, values,
            scale=1 / math.sqrt(self.head_dim),
            mask=None if length == 1 else "causal",
        )
        return self.o_proj(
            attended.transpose(0, 2, 1, 3).reshape(1, length, self.heads * self.head_dim)
        )


class Qwen3Block(nn.Module):
    def __init__(self, config: Qwen3Config) -> None:
        super().__init__()
        self.self_attn = Qwen3Attention(config)
        self.mlp = SwiGLU(config.hidden_size, config.intermediate_size)
        self.input_layernorm = nn.RMSNorm(config.hidden_size, eps=config.rms_norm_eps)
        self.post_attention_layernorm = nn.RMSNorm(config.hidden_size, eps=config.rms_norm_eps)

    def __call__(self, x: mx.array, cache: KVCache) -> mx.array:
        attended = x + self.self_attn(self.input_layernorm(x), cache)
        return attended + self.mlp(self.post_attention_layernorm(attended))


class Qwen3Trunk(nn.Module):
    def __init__(self, config: Qwen3Config) -> None:
        super().__init__()
        self.embed_tokens = nn.Embedding(config.vocab_size, config.hidden_size)
        self.layers = [Qwen3Block(config) for _ in range(config.num_hidden_layers)]
        self.norm = nn.RMSNorm(config.hidden_size, eps=config.rms_norm_eps)


class Qwen3Activations(NamedTuple):
    embeddings: mx.array
    blocks: list[mx.array]
    norm: mx.array
    logits: mx.array


class Qwen3(nn.Module):
    def __init__(self, config: Qwen3Config) -> None:
        super().__init__()
        self.config = config
        self.model = Qwen3Trunk(config)
        if not config.tie_word_embeddings:
            self.lm_head = nn.Linear(config.hidden_size, config.vocab_size, bias=False)

    def make_cache(self) -> list[KVCache]:
        return [KVCache() for _ in self.model.layers]

    def head(self, normed: mx.array) -> mx.array:
        if self.config.tie_word_embeddings:
            return self.model.embed_tokens.as_linear(normed)
        return self.lm_head(normed)

    def activations(self, ids: mx.array, cache: list[KVCache] | None = None) -> Qwen3Activations:
        cache = cache if cache is not None else self.make_cache()
        x = self.model.embed_tokens(ids)
        embeddings = x
        blocks: list[mx.array] = []
        for block, layer_cache in zip(self.model.layers, cache, strict=True):
            x = block(x, layer_cache)
            blocks.append(x)
        normed = self.model.norm(x)
        return Qwen3Activations(embeddings, blocks, normed, self.head(normed))

    def __call__(self, ids: mx.array, cache: list[KVCache] | None = None) -> mx.array:
        return self.activations(ids, cache).logits
